### PR TITLE
Keyboard navigation for search

### PR DESCRIPTION
**Pull request by microbit-robert**
Adds keyboard navigation for search results using tab, as well as up/down arrows. Hover/focus styles added, and viewed search results are tracked and styled appropriately. 

See #440

### CODE REVIEW

File: src/documentation/search/SearchDialog.tsx [hunk 1]
```diff
@@ -42,6 +46,47 @@ const SearchDialog = ({
...
+        }
+      }
+    };
+
+    document.addEventListener("keydown", keydown);
+    return () => {
+      document.removeEventListener("keydown", keydown);
+    };
+  }, []);
+
```
Comment (microbit-robert) [inline src/documentation/search/SearchDialog.tsx:89]: @microbit-matt-hillsdon This is horrible. This overrides the default behaviour of being able to scroll through the search results using the up and down arrows, which on reflection, is not something I'm sure we want to do? If we do want this, perhaps there is a better way?

Comment (microbit-matt-hillsdon) [reply src/documentation/search/SearchDialog.tsx:89]: Yeah, worth some further thought. Everything else here is more straightforwardly valuable (and fixed tabbing) so perhaps split this bit out and let's get the rest in?

File: src/documentation/search/SearchResultList.tsx [hunk 2]
```diff
@@ -38,37 +48,51 @@ const SearchResultList = ({
...
+    <Stack pl="3px" pr="3px">
+      <Link
+        bgColor={viewedResults.includes(id) ? "#efedf5" : "unset"}
+        borderRadius="md"
+        href={url}
+        onClick={(e) => {
+          e.preventDefault();
+          onClose();
+          setState(navigation);
+          onViewResult(id);
```
Comment (microbit-matt-hillsdon) [inline src/documentation/search/SearchResultList.tsx:76]: Maybe the setState call should move up to inside the handler passed for onViewResult?

File: src/router-hooks.tsx [hunk 3]
```diff
@@ -99,14 +110,7 @@ export const RouterProvider = ({ children }: { children: ReactNode }) => {
...
-        .filter(([_, v]) => !!v)
-        .map(([k, v]) => {
-          return `${encodeURIComponent(k)}=${encodeURIComponent(
-            serializeValue(v)
-          )}`;
-        })
-        .join("&");
+      const query = getQueryString(newState);
       const url =
         window.location.toString().split("?")[0] + (query ? "?" + query : "");
```
Comment (microbit-matt-hillsdon) [inline src/router-hooks.tsx:115]: Is this bit common between the new usage and old? Perhaps a `toUrl(state: RouterState)` API would be better and fit both usecases?

**Review by microbit-matt-hillsdon — COMMENTED** [review]
I really like the highlighting, it feels much more useful than I was expecting.

Added a couple of minor comments.

File: src/documentation/search/SearchResultList.tsx (hunk 2 above)
Comment (microbit-robert) [reply src/documentation/search/SearchResultList.tsx]: [`f2e178f`](https://github.com/microbit-foundation/python-editor-next/pull/477/commits/f2e178f02edf9d9bf73e5c8a243016d57fb74dba). We now handle closing the on in handleViewResult as well, so this no longer has to be passed down.

File: src/router-hooks.tsx (hunk 3 above)
Comment (microbit-robert) [reply src/router-hooks.tsx]: [`1fe7d6e`](https://github.com/microbit-foundation/python-editor-next/pull/477/commits/1fe7d6e76710879f26bd60a083836514a42eb880)

File: src/documentation/search/SearchResultList.tsx [hunk 4]
```diff
@@ -37,38 +44,45 @@ const SearchResultList = ({
...
 
 const SearchResultItem = ({
-  value: { extract, navigation, containerTitle, title },
-  onClose,
+  value: { extract, navigation, containerTitle, title, id },
+  viewedResults,
+  onViewResult,
 }: SearchResultItemProps) => {
-  const [, setState] = useRouterState();
+  const url = toUrl(getQueryString(navigation), navigation);
```
Comment (microbit-matt-hillsdon) [inline src/documentation/search/SearchResultList.tsx:56]: Let's move the getQueryString call inside toUrl so the caller just gets a URL from the router state and doesn't have to make two calls.

Comment (microbit-robert) [reply src/documentation/search/SearchResultList.tsx]: [`a905f61`](https://github.com/microbit-foundation/python-editor-next/pull/477/commits/a905f6178a9918ecaed74cf31ef8b850b8005960)

Comment (microbit-matt-hillsdon) [reply src/documentation/search/SearchResultList.tsx]: Change looks great, will just wait for it to build then merge.

**Review by microbit-matt-hillsdon — APPROVED** [review]
Looks great.

**Pull request merged**